Documentation/proposals/202411-zone-aware-sharding.md

```diff
@@ -0,0 +1,357 @@
+
+```text
+0 1 2 | 3 4 5 | 6 7 8 | 9 | shard index
+A B C | A B C | A B C | A | zone
+0 0 0 | 1 1 1 | 2 2 2 | 0 | assignment index
+```
+
+In this case the 2nd assert will warn about double scraping of instances in
+zone A, as the same targets are being assigned to instance 0 and 9.
+
+When `num_shards` is 2 and `len(zones)` is 3 as in `[A..C]`.
+`shards_per_zone` is 1. This yields the following distribution
+
+```text
+0 1   | shard index
+A B C | zone
+0 0 0 | assignment index
+```
+
+In this case targets in zone C are not being scraped.
+
+Both cases should lead to an error during reconciliation, causing the change to
+not be rolled out. The first case (double scraping) is not as severe as a zone
+not being scraped but it is otherwise hard to spot.
+It's also to be mentioned that replicas are to be used to achieve redundant
+scraping.
+
+### Topology field discovery
+
+The kubernetes service discovery currently does not expose any topology field.
+Such a field would have to be added, otherwise users would have to inject such
+a label themselves.
+
+A good candidate for such fields are the `topology.kubernetes.io/*` labels
+which should be present on all nodes.
+
+There are two ways to handle this:
+
+1. A change to the prometheus kubernetes discovery service to add the required
+   label to all targets.
+2. The operator could do this discovery and add a relabel rule based on the
+   node name.
+
+The second solution would require the operator to constantly update the relabel
+configuration. This could lead to increased load on clusters with agressive
+autoscaling as well as race conditions for pods on newly created nodes, as the
+config change is not atomic/instant.
+
+As of that, a change to the kubernetes service discovery is considered the more
+stable, and thus preferrable solution. It will require additional permissions
+for prometheus in case it is not already allowed to read node objects.
+
+### API changes
+
+> [!NOTE]
+> This proposal is mutually exclusive to [DaemonSet mode](./202405-agent-daemonset.md),
+> as prometheus always scrapes a single node in that case.
+> Defining a `shardingStrategy` when `DaemonSet mode` is active, should lead to
+> a reconciliation error.
+
+Following the algorithm presented above, we suggest the following configuration
+options to be added to the [Prometheus](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#monitoring.coreos.com/v1.Prometheus) and PrometheusAgent custom resource definitions.
+
+All values used in this snippet should also be the defaults for their
+corresponding keys.
+
+```yaml
+spec:
+  shardingStrategy:
+     # Select a sharding mode. Can be 'Classic' or 'Topology'.
+     # Defaults to `Classic`.
+    mode: 'Classic'    
+
+    # The following section is only valid if "mode" is set to "Topology"
+    topology:
+        # All topology values to be used by the cluster, i.e. a list of all
+        # zones in use.
+        values: []
```

Comment:
saying it here to remember later but it doesn't have to be in the proposal: if the mode is Topology then the values list should have at least 1 item.